### DISCUSSION TITLE
Is vcpkg is data-race free?

### DISCUSSION THREAD

**Discussion by ixSci** [microsoft/vcpkg#42056] [Q&A] [answered]
Suppose I have a vcpkg installed on some machine and 2 jobs get started simultaneously: clang & gcc builds of some project utilizing vcpkg. If the packages can't be acquired from the cache, they have to be built and cached. So what happens when both jobs tries to download/build/cache the packages? Does vcpkg have some internal synchronization which handles this situation?

**Comment by autoantwort** [comment, marked as answer]
There is internal synchronization (a lock file). If you use `--x-wait-for-lock` the instances wait until the other instances are done.

**Reply by ixSci** [reply]
> There is internal synchronization (a lock file). If you use `--x-wait-for-lock` the instances wait until the other instances are done.

So there is a lock file when I use the `--x-wait-for-lock` command parameter? Can't find any mention of this parameter in the docs (only its usage), is it some arcane non-public parameter? :)

Also, how do I use it from within CMake integration?

**Reply by autoantwort** [reply]
There is always a lock file. Usually you get an error when some other instance is already running. 

> Also, how do I use it from within CMake integration?

The cmake integration does this by default

**Reply by ixSci** [reply]
> There is always a lock file. Usually you get an error when some other instance is already running.

What is the point of this parameter, then?

**Reply by autoantwort** [reply]
Like the name suggests then vcpkg waits until the lock is released instead of printing an error and exit.

**Reply by ixSci** [reply]
> Like the name suggests then vcpkg waits until the lock is released instead of printing an error and exit.

Ah, ok. So the default behavior is to bail out and with the parameter we can force other instances to get in line. Got it. Thank you!

**Reply by autoantwort** [reply]
Ok it seems that this only works in manifest mode (I have never used classic mode)
